File: src/lsdsoft/welltools/im/immn73/IMMN73DataFilter.cpp
#include <math.h>
#include "IMMN73DataFilter.hpp"

#define IN_BUFFER_SIZE 4

static bool allowedParams[] = { false, false, true, false, true, true, false, false };

void IMMN73DataFilter::connect() {
  port.openPort(portID);
  port.setSerialPortParams(2400,
      SerialPort::DATABITS_8,
      SerialPort::STOPBITS_1,
      SerialPort::PARITY_ODD);
  port.setInBufferSize(IN_BUFFER_SIZE);
  port.setOutBufferSize(1);
  port.setDTR(true);
  port.setRTS(false);
  port.monitor(true);
  connected = true;
}

void IMMN73DataFilter::ask(unsigned int parameter, bool singleRequest) {
char buffer[3];
  if(parameter > 7) {
    nextParam = -1;
    return;
  }
  //if(parameter == 2) Sleep(1000);
  buffer[0] = 'S';
  buffer[1] = '0' + parameter;
  buffer[2] = 13;
  askedParam = parameter;
  startPacket = true;
  port.clearInput();
  port.Write(buffer, 3);
  arrived = false;
  if(singleRequest)
    nextParam = -1;
  else
  switch(parameter) {
    case 2: nextParam = 4; break;
    case 4: nextParam = 5; break;
    case 5: nextParam = 2; break;
  }
  //nextParam = singleRequest?-1:parameter + 1;
}

void IMMN73DataFilter::reset() {
char * buffer = "R\x0d";
  port.clearInput();
  port.Write(buffer, 2);
}

void IMMN73DataFilter::ask() {
  // запрос нулей датчиков
  ask(2, false);
//  waitForArrived();
  // запрос значения коордынаты X
//  ask(4);
//  waitForArrived();
  // запрос значения координаты Y

//  ask(5);
//  waitForArrived();
}

int IMMN73DataFilter::hexCharToInt(char hex) {
 if(hex >= '0' && hex <= '9') return (hex - '0');
 if(hex >= 'A' && hex <= 'F') return (hex - 'A' + 10);
 return -1;
}

int IMMN73DataFilter::decodeChunk(unsigned char * buffer) {
// arrived string looks like this: "F7F8"
// but signful values only las three
  int mul = 1;
  int ret = 0;
  for(int i = 0; i < 3; i++) {
    int value = hexCharToInt(buffer[3 - i]);

    ret += value * mul;
    mul *= 16;
  }
  return ret;
}

void IMMN73DataFilter::waitForArrived() {
  while(!arrived)Sleep(50);
}

IMMN73DataFilter::IMMN73DataFilter() : fifo(256), ChannelDataSource(16){
//  port = 0;
  //eventListener = 0;
  //dataEventListener = 0;
  packetCount = 0;
  //isSync = false;
  inputBuffer = new unsigned char[IN_BUFFER_SIZE];
  bufferIndex = 0;
  port.addEventListener(*this);
  //id = "immn73";
  //sem = CreateSemaphore(NULL, 0, 1, "sem\ion1");
}

//IONDataSource::IONDataSource(SerialPort* sPort) {
//  IONDataSource::IONDataSource();
//  setPort(sPort);
//}

IMMN73DataFilter::~IMMN73DataFilter() {
  port.closePort();
  delete[] inputBuffer;
}

//---------------------------------------------------------------------------
void IMMN73DataFilter::getInput() {
char buf[IN_BUFFER_SIZE];
  if(startPacket) {
    fifo.clear();
    startPacket = false;
  }
  //port->Read(inputBuffer, IN_BUFFER_SIZE);
  port.Read(buf, IN_BUFFER_SIZE);
  for(int i = 0; i < IN_BUFFER_SIZE; i++)
    if(hexCharToInt(buf[i]) >= 0)
      fifo.push(buf[i]);
    //fifo.push(0);
//  bufferIndex = 0;
}
//---------------------------------------------------------------------------
bool IMMN73DataFilter::findPacket() {
  if(fifo.count() < IMMN73_PACKET_SIZE) return false;
  for(int i = 0; i < IMMN73_PACKET_SIZE; i++) {
    packetBuffer[i] = fifo.pop();
  }
  return true;
}
//---------------------------------------------------------------------------
void IMMN73DataFilter::processPacket() {
unsigned int * packet = (unsigned int *)&packetBuffer;
  for(int i = 0; i < 2; i++) {
    int value = decodeChunk(packetBuffer + i * 4);
    ChannelDataEvent ev(askedParam * 2 + i, value);
    sendEvent(ev);
  }
  arrived = true;
  if(nextParam == 2) {
    //reset();
    //Sleep(500);
  }

  if(nextParam >=0) {
    ask(nextParam, false);
    Sleep(400);
    reset();
  }

}
//---------------------------------------------------------------------------
void IMMN73DataFilter::serialEvent(SerialPortEvent& ev) {
//  if((eventListener == NULL) && (dataEventListener == NULL)) {
//    port->clearInput();
//    return;
//  }
int e = ev.getEventType();
  switch(e) {
    case SerialPortEvent::DATA_AVAILABLE:
      getInput();
//      findPacket();
//      processPacket();
      while(findPacket()) {
        processPacket();
      }
      //port->clearInput();
      break;
  }
}

File: src/lsdsoft/welltools/im/immn73/IMMN73DataFilter.java
///////////////////////////////////////////////////////////////////////
//
// (c) Copyright 2003 "Ural-Geo"
//     info@uralgeo.com
//
// author: lsdsoft@mail.ru
//
///////////////////////////////////////////////////////////////////////
/*
  Фильтр данных для скважинного инклинометра ИММН 73
  Разделяет сплошной поток из последовательного порта, посылаемого панелью,
  на каналы показаний датчиков. После принятия полного пакета вызывает обработчик каналов

*/
package lsdsoft.welltools.im.immn73;

import lsdsoft.metrolog.*;
import com.lsdsoft.comm.*;
import javax.comm.*;
import java.io.*;

//#include "ChannelDataSource.hpp"
//#include "SerialConnection.hpp"
//#include "SerialPortEventListener.hpp"
//#include "IMDataEventListener.hpp"
//#include "ComPort.h"
//#include "FIFOBuffer.hpp"

//#define IMMN73_PACKET_SIZE 8
//#define IONCHANNELS 15

public class IMMN73DataFilter extends ChannelDataSource
               implements SerialPortEventListener
                     {
  private static final int IMMN73_PACKET_SIZE = 8;
  private static final int COMMAND_SELECT = 1;
  private static final int COMMAND_REQUEST = 2;
  private CommConnection connection = new CommConnection();
  private SerialPort port;
  private InputStream ins = null;
  private OutputStream outs = null;
  private byte[] inputBuffer = new byte[IMMN73_PACKET_SIZE];
  private byte[] requestPacket = {0x52, 0x0d};
  private byte[] selectPacket = {0x53, 0x30, 0x0d};
  private int bufferIndex;
  private int command;
  private boolean commandDone = false;
  private int askedSensor;
  //bool flagBufferAsData;
  byte[] packetBuffer;
  //FIFOBuffer fifo;
  int askedParam;
  private boolean startPacket;
  int packetCount;
  int nextParam;
  //InklinometerAngles angles;
  // wait and read packet (3 bytes)
//  void getPacket();
  // bytes to skip from buffer
  //int toSkip;
  //bool isSync;
  //boolean arrived;
  //HANDLE sem;
  /**
   * Initializing serial port for IMMN73 panel
   */
  private void setupPort() {
    try {
    port.setSerialPortParams(2400,
        SerialPort.DATABITS_8,
        SerialPort.STOPBITS_1,
        SerialPort.PARITY_ODD);
    //port.setInBufferSize(IMMN73_PACKET_SIZE);
    //port.setOutBufferSize(32);
    }
    catch (Exception ex) {}
  }
  /**
   *
   */
  public void setPortName(String portName) {
    connection.setPortName(portName);
  }
  /**
   *
   */
  public void selectSensor(int sensor) {
    if(sensor > 7) {
      return;
    }
    selectPacket[1] = (byte)(0x30 + sensor);
    askedSensor = sensor;
    try {
      outs.flush();
      outs.write(selectPacket, 0, 3);
      command = COMMAND_SELECT;
      commandDone = false;
    } catch(Exception ex) {
      System.err.print(ex.getMessage());
    }
  }
  /**
   *
   */
  public void request() throws Exception{
    command = COMMAND_REQUEST;
    commandDone = false;
    outs.write(requestPacket);
  }
//  void sync();
//  void skip();
  /**
   *
   */
  private void getInput() throws Exception{
    ins.read(packetBuffer, 0, IMMN73_PACKET_SIZE);
  }
//  public boolean findPacket();
//  int decodeChunk(unsigned char * buffer);
  /**
   *
   */
  void processPacket() {
    if(command == COMMAND_SELECT) {
      commandDone = true;
    }
      //port.
      /*
    COMMAND_READ: break;
    case COMMAND_SELECT:
       commandDone = true;
       port.clearInput();
       Sleep(150);
       reset();
//     Sleep(750);
//     reset();
       return;
    }
    //if(!reseted) {

    port.clearInput();
    for(int i = 0; i < 2; i++) {
      int value = decodeChunk(packetBuffer + i * 4);
      ChannelDataEvent ev(askedParam * 2 + i, value);
      sendEvent(ev);
    }
    arrived = true;
    commandDone = true;
*/
  }
  void waitForArrived() {

  }
  int hexCharToInt(char hex) {
    int ret = 0;
    return ret;
  }

  public IMMN73DataFilter(String portName) {
    setPortName(portName);

  }
  public CommConnection getConnection() {
    return connection;
  }
  /**
   * Serial port event handler
   */
  public void serialEvent(SerialPortEvent ev) {
    switch(ev.getEventType()) {
      case SerialPortEvent.DATA_AVAILABLE:
        try {
          getInput();
          processPacket();
        } catch(Exception ex) {
          System.err.print(ex.getMessage());
        }
        break;
    }
  }
  /**
   *
   */
  public void connect() {
    try {
      CommPort cport;
      connection.connect();
      cport = connection.getPort();
      if(cport instanceof SerialPort) {
        port = (SerialPort)cport;
        setupPort();
        port.addEventListener(this);
        ins = port.getInputStream();
        outs = port.getOutputStream();
      } else
        throw new Exception("Invalid communication port type: " + connection.getPortName());
    } catch(Exception ex) {
      System.err.print(ex.getMessage());
    }
  }
  public void disconnect() {
    if(port != null) {
      connection.disconnect();
      port = null;
    }
  }
  //void addEventListener(ChannelDataEventListener* evListener);
  //void removeEventListener();
  void ask() {
  }

//  void setBufferAsData(bool state);
//  void measureAverage(unsigned int count, InklinometerAngles& angles);
  // process one packet
//  void process();
  private void finish() {
  }
  //int getChannel() { return channel; }
  //int getChannelValue(int chanIndex) { return channelData[chanIndex]; }
  //int getValue() {return value; }
  //void setPort(SerialPort * port);
  //virtual void getAngles(InklinometerAngles &ang);
  //void getAngles(double & rotate, double& zenit, double& azimut);


};


